File: src/com/ptby/dynamicreturntypeplugin/scanner/MethodCallReturnTypeScanner.java
package com.ptby.dynamicreturntypeplugin.scanner;

import com.jetbrains.php.lang.psi.elements.MethodReference;
import com.ptby.dynamicreturntypeplugin.config.ClassMethodConfigKt;
import com.ptby.dynamicreturntypeplugin.gettype.GetTypeResponse;
import com.ptby.dynamicreturntypeplugin.typecalculation.CallReturnTypeCalculator;

import java.util.List;

public class MethodCallReturnTypeScanner {

    private final CallReturnTypeCalculator callReturnTypeCalculator;


    public MethodCallReturnTypeScanner( CallReturnTypeCalculator callReturnTypeCalculator ) {
        this.callReturnTypeCalculator = callReturnTypeCalculator;
    }


    public GetTypeResponse getTypeFromMethodCall( List<ClassMethodConfigKt> classMethodConfigList,
                                                  MethodReference  methodReference ) {
        for ( ClassMethodConfigKt classMethodConfig : classMethodConfigList ) {
            if ( classMethodConfig.equalsMethodReferenceName( methodReference ) ) {
                GetTypeResponse getTypeResponse = callReturnTypeCalculator
                        .calculateTypeFromMethodParameter( methodReference, classMethodConfig.getParameterIndex() );
                if ( !getTypeResponse.isNull() ) {
                    return getTypeResponse;
                }
            }
        }

        return new GetTypeResponse( null );
    }


}

File: src/com/ptby/dynamicreturntypeplugin/scanner/MethodCallReturnTypeScanner.kt
package com.ptby.dynamicreturntypeplugin.scanner

import com.jetbrains.php.lang.psi.elements.ArrayAccessExpression
import com.jetbrains.php.lang.psi.elements.MethodReference
import com.ptby.dynamicreturntypeplugin.config.ClassMethodConfigKt
import com.ptby.dynamicreturntypeplugin.gettype.ArrayAccessGetTypeResponse
import com.ptby.dynamicreturntypeplugin.gettype.FunctionReferenceGetTypeResponse
import com.ptby.dynamicreturntypeplugin.gettype.GetTypeResponse

public class MethodCallReturnTypeScanner() {


    public fun getTypeFromMethodCall(classMethodConfigList: List<ClassMethodConfigKt>,
                                     methodReference: MethodReference): FunctionReferenceGetTypeResponse {

        for (classMethodConfig in classMethodConfigList) {
            if (classMethodConfig.equalsMethodReferenceName(methodReference)) {
                return FunctionReferenceGetTypeResponse.newMethod(methodReference)
            }
        }

        return FunctionReferenceGetTypeResponse.createNull()
    }


    public fun getTypeFromArrayAccess(arrayAccessExpression: ArrayAccessExpression): GetTypeResponse {

        return ArrayAccessGetTypeResponse.newArrayAccess(arrayAccessExpression)
    }


}
